// @ionic/cli-utils/lib/ionic-angular/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk = require("chalk");
const utils_1 = require("./utils");
function build({ env, options }) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const appScriptsArgs = yield buildOptionsToAppScriptsArgs(options);
        process.argv = ['node', 'appscripts'].concat(appScriptsArgs);
        const AppScripts = yield utils_1.importAppScripts(env);
        const context = AppScripts.generateContext();
        env.log.info(`Running app-scripts build: ${chalk.bold(appScriptsArgs.join(' '))}\n`);
        return yield AppScripts.build(context);
    });
}
exports.build = build;
function buildOptionsToAppScriptsArgs(options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { minimistOptionsToArray } = yield Promise.resolve().then(function () { return require('../utils/command'); });
        const minimistArgs = {
            _: [],
            prod: options.prod ? true : false,
            aot: options.aot ? true : false,
            minifyjs: options.minifyjs ? true : false,
            minifycss: options.minifycss ? true : false,
            optimizejs: options.optimizejs ? true : false,
            platform: options.platform,
            target: options.target,
        };
        return minimistOptionsToArray(minimistArgs, { useEquals: false });
    });
}
exports.buildOptionsToAppScriptsArgs = buildOptionsToAppScriptsArgs;
