// ionic/dist/commands/git/remote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk = require("chalk");
const cli_utils_1 = require("@ionic/cli-utils");
const command_1 = require("@ionic/cli-utils/lib/command");
const fs_1 = require("@ionic/cli-utils/lib/utils/fs");
let GitRemoteCommand = class GitRemoteCommand extends command_1.Command {
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { App } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/app'); });
            const token = yield this.env.session.getUserToken();
            const appId = yield this.env.project.loadAppId();
            const appLoader = new App(token, this.env.client);
            const app = yield appLoader.load(appId);
            if (!app.repo_url) {
                throw this.exit(`Missing ${chalk.bold('repo_url')} property in app.`);
            }
            if (!(yield fs_1.pathExists(path.join(this.env.project.directory, '.git')))) {
                this.env.log.warn(`Initializing a git repository for your project.\n` +
                    `Before your first ${chalk.green('git push ionic master')}, you'll want to commit all the files in your project:\n\n` +
                    `${chalk.green('git commit -a -m "Initial commit"')}\n`);
                yield this.env.shell.run('git', ['init'], { showSpinner: false, cwd: this.env.project.directory });
            }
            const remote = app.repo_url;
            const regex = /ionic\t(.+) \(\w+\)/;
            // would like to use get-url, but not available in git 2.0.0
            const remotes = yield this.env.shell.run('git', ['remote', '-v'], { showCommand: false, cwd: this.env.project.directory });
            let found = false;
            let matches = true;
            for (let line of remotes.split('\n')) {
                const match = regex.exec(line.trim());
                if (match) {
                    found = true;
                    if (match[1] !== remote) {
                        matches = false;
                        break;
                    }
                }
            }
            if (found) {
                if (matches) {
                    this.env.log.info(`Existing remote ${chalk.bold('ionic')} found.`);
                }
                else {
                    yield this.env.shell.run('git', ['remote', 'set-url', 'ionic', remote], { showSpinner: false, cwd: this.env.project.directory });
                    this.env.log.ok(`Updated remote ${chalk.bold('ionic')}.`);
                }
            }
            else {
                yield this.env.shell.run('git', ['remote', 'add', 'ionic', remote], { showSpinner: false, cwd: this.env.project.directory });
                this.env.log.ok(`Added remote ${chalk.bold('ionic')}.`);
            }
        });
    }
};
GitRemoteCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'remote',
        type: 'project',
        backends: [cli_utils_1.BACKEND_PRO],
        description: 'Adds/updates the Ionic git remote to your local Ionic app repository',
    })
], GitRemoteCommand);
exports.GitRemoteCommand = GitRemoteCommand;
