// ionic/dist/commands/cordova/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
const fs_1 = require("@ionic/cli-utils/lib/utils/fs");
const serve_1 = require("@ionic/cli-utils/lib/serve");
exports.CORDOVA_RUN_COMMAND_OPTIONS = [
    {
        name: 'list',
        description: 'List all available Cordova targets',
        type: Boolean,
        intent: 'CORDOVA',
    },
    {
        name: 'livereload',
        description: 'Spin up server to live-reload www files',
        type: Boolean,
        aliases: ['l'],
    },
    {
        name: 'consolelogs',
        description: 'Print out console logs to terminal',
        type: Boolean,
        aliases: ['c'],
    },
    {
        name: 'serverlogs',
        description: 'Print out dev server logs to terminal',
        type: Boolean,
        aliases: ['s'],
    },
    {
        name: 'address',
        description: 'Use specific address for dev/live-reload server',
        default: serve_1.BIND_ALL_ADDRESS,
    },
    {
        name: 'port',
        description: 'Use specific port for the dev server',
        default: String(serve_1.DEFAULT_SERVER_PORT),
        aliases: ['p'],
    },
    {
        name: 'livereload-port',
        description: 'Use specific port for live-reload server',
        default: String(serve_1.DEFAULT_LIVERELOAD_PORT),
        aliases: ['r'],
    },
    {
        name: 'prod',
        description: 'Mark as a production build',
        type: Boolean,
    },
    {
        name: 'aot',
        description: 'Perform ahead-of-time compilation for this build',
        type: Boolean,
    },
    {
        name: 'minifyjs',
        description: 'Minify JS for this build',
        type: Boolean,
    },
    {
        name: 'minifycss',
        description: 'Minify CSS for this build',
        type: Boolean,
    },
    {
        name: 'optimizejs',
        description: 'Perform JS optimizations for this build',
        type: Boolean,
    },
    {
        name: 'debug',
        description: 'Mark as a debug build',
        type: Boolean,
        intent: 'CORDOVA',
    },
    {
        name: 'release',
        description: 'Mark as a release build',
        type: Boolean,
        intent: 'CORDOVA',
    },
    {
        name: 'device',
        description: 'Deploy Cordova build to a device',
        type: Boolean,
        intent: 'CORDOVA',
    },
    {
        name: 'emulator',
        description: 'Deploy Cordova build to an emulator',
        type: Boolean,
        intent: 'CORDOVA',
    },
    {
        name: 'target',
        description: `Deploy Cordova build to a device (use ${chalk.green('--list')} to see all)`,
        type: String,
        intent: 'CORDOVA',
    },
    {
        name: 'buildConfig',
        description: 'Use the specified Cordova build configuration',
        intent: 'CORDOVA',
    },
];
class CordovaCommand extends command_1.Command {
    preRunChecks() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { prettyPath } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/utils/format'); });
            const { checkCordova } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/utils'); });
            yield checkCordova(this.env);
            // Check for www folder
            if (this.env.project.directory) {
                const wwwPath = path.join(this.env.project.directory, 'www');
                const wwwExists = yield fs_1.pathExists(wwwPath); // TODO: hard-coded
                if (!wwwExists) {
                    this.env.tasks.next(`Creating ${chalk.bold(prettyPath(wwwPath))} directory for you`);
                    yield fs_1.fsMkdir(wwwPath, undefined);
                    this.env.tasks.end();
                }
            }
        });
    }
    runCordova(argList, _a = {}) {
        var { fatalOnNotFound = false, truncateErrorOutput = 5000 } = _a, options = tslib_1.__rest(_a, ["fatalOnNotFound", "truncateErrorOutput"]);
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { ERROR_SHELL_COMMAND_NOT_FOUND } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/shell'); });
            const { pkgManagerArgs } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/utils/npm'); });
            try {
                return yield this.env.shell.run('cordova', argList, Object.assign({ fatalOnNotFound, truncateErrorOutput }, options));
            }
            catch (e) {
                if (e === ERROR_SHELL_COMMAND_NOT_FOUND) {
                    const cdvInstallArgs = yield pkgManagerArgs(this.env, { pkg: 'cordova', global: true });
                    throw this.exit(`The Cordova CLI was not found on your PATH. Please install Cordova globally:\n\n` +
                        `${chalk.green(cdvInstallArgs.join(' '))}\n`);
                }
                if (options.fatalOnError) {
                    this.env.log.nl();
                    this.env.log.error('Cordova encountered an error.\nYou may get more insight by running the Cordova command above directly.\n');
                }
                throw e;
            }
        });
    }
    checkForPlatformInstallation(runPlatform) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (runPlatform) {
                const { ConfigXml } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/config'); });
                const conf = yield ConfigXml.load(this.env.project.directory);
                const platforms = yield conf.getPlatformEngines();
                if (!platforms.map(p => p.name).includes(runPlatform)) {
                    const { installPlatform } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/project'); });
                    yield installPlatform(this.env, runPlatform);
                }
            }
        });
    }
}
exports.CordovaCommand = CordovaCommand;
class CordovaRunCommand extends CordovaCommand {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { filterArgumentsForCordova } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/utils'); });
            yield this.preRunChecks();
            if (options['list']) {
                const args = filterArgumentsForCordova(this.metadata, inputs, options);
                if (!options['device'] && !options['emulator']) {
                    if (args[0] === 'run') {
                        args.push('--device');
                    }
                    else if (args[0] === 'emulate') {
                        args.push('--emulator');
                    }
                }
                args[0] = 'run';
                yield this.runCordova(args, { showExecution: true });
                return 0;
            }
            if (!inputs[0]) {
                const platform = yield this.env.prompt({
                    type: 'input',
                    name: 'platform',
                    message: `What platform would you like to run: ${chalk.green('ios')}, ${chalk.green('android')}:`,
                });
                inputs[0] = platform.trim();
            }
            yield this.checkForPlatformInstallation(inputs[0]);
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { ConfigXml } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/config'); });
            const { filterArgumentsForCordova, generateBuildOptions } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/utils'); });
            if (!options['livereload'] && (options['consolelogs'] || options['serverlogs'])) {
                this.env.log.info(`${chalk.green('--consolelogs')} or ${chalk.green('--serverlogs')} detected, using ${chalk.green('--livereload')}`);
                options['livereload'] = true;
            }
            const isLiveReload = options['livereload'];
            const conf = yield ConfigXml.load(this.env.project.directory);
            yield conf.resetContentSrc();
            yield conf.save();
            if (isLiveReload) {
                const { serve } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/serve'); });
                const serverDetails = yield serve(this.env, inputs, generateBuildOptions(this.metadata, options));
                if (serverDetails.externallyAccessible === false) {
                    const extra = serve_1.LOCAL_ADDRESSES.includes(serverDetails.externalAddress) ? '\nEnsure you have proper port forwarding setup from your device to your computer.' : '';
                    this.env.log.warn(`Your device or emulator may not be able to access ${chalk.bold(serverDetails.externalAddress)}.${extra}\n\n`);
                }
                yield conf.writeContentSrc(`${serverDetails.protocol || 'http'}://${serverDetails.externalAddress}:${serverDetails.port}`);
                yield conf.save();
            }
            else {
                const { build } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/build'); });
                yield build(this.env, inputs, generateBuildOptions(this.metadata, options));
            }
            yield this.runCordova(filterArgumentsForCordova(this.metadata, inputs, options), { showExecution: true });
            if (!isLiveReload) {
                this.env.log.ok(`Your app has been deployed.\n` +
                    `Did you know you can live-reload changes from your app with ${chalk.green('--livereload')}?`);
            }
        });
    }
}
exports.CordovaRunCommand = CordovaRunCommand;
